### ISSUE TITLE
Printer is not recognized when SANE is enabled

### ISSUE THREAD

**Issue by maydayv7** [NixOS/nixpkgs#147217]
### Describe the bug
When `hardware.sane` is enabled, my printer is not recognized by `cups`. Initially, when the option was not enabled, my printer would immediately get recognized upon plugging in the USB, show up in the Printers section in `gnome-control-center`, and printing also works fine. After enabling SANE, the printer is not recognized, but the in-built scanner gets recognized and works properly in apps like `simple-scan`. After disabling it and running `nixos-rebuild switch`, the printer is again immediately recognized, but obviously the scanner doesn't work.

### Steps To Reproduce
Configuration:
```
hardware.sane.enable = true;
services.printing =
{
  enable = true;
  drivers = with pkgs; [ gutenprint cnijfilter2 ];
};
```

### Expected behavior
Both the printer and scanner must function and get recognized properly when the above-mentioned configuration is applied.

### Logs
*SANE did not give any errors*
`journalctl -xb | grep cups`
Error after enabling `hardware.sane`:
```
Nov 23 21:40:31 Vortex cupsd[1315]: [Client 17] Started "/nix/store/qdnc2arp8pqjv1y9989lr9rr657iclsf-cups-progs/lib/cups/daemon/cups-deviced" (pid=3347, file=16)
Nov 23 21:40:31 Vortex cupsd[1315]: [CGI] cups-brf must be called as root
Nov 23 21:40:31 Vortex cupsd[1315]: [cups-deviced] PID 3349 (cups-brf) stopped with status 1!
Nov 23 21:40:32 Vortex cupsd[1315]: Expiring subscriptions...
Nov 23 21:40:33 Vortex cupsd[1315]: [cups-driverd] Read "/var/cache/cups/ppds.dat", 148 PPDs...
Nov 23 21:40:33 Vortex cupsd[1315]: [cups-driverd] No new or changed PPDs...
```

After disabling it, the printer gets recognized:
```
Nov 23 22:05:21 Vortex cupsd[1315]: REQUEST localhost - - "POST / HTTP/1.1" 200 175 CUPS-Get-Devices -
Nov 23 22:05:22 Vortex cupsd[1315]: Expiring subscriptions...
Nov 23 22:05:22 Vortex cupsd[1315]: REQUEST localhost - - "POST / HTTP/1.1" 200 1027377 CUPS-Get-PPDs -
Nov 23 22:05:22 Vortex cupsd[1315]: REQUEST localhost - - "POST /admin/ HTTP/1.1" 401 345 CUPS-Add-Modify-Printer successful-ok
Nov 23 22:05:22 Vortex cupsd[1315]: Setting MG2500-series device-uri to "usb://Canon/MG2500%20series?serial=16A80E&interface=1" (was "file:///dev/null".)
Nov 23 22:05:22 Vortex cupsd[1315]: Registering ICC color profiles for "MG2500-series".
Nov 23 22:05:22 Vortex cupsd[1315]: New printer "MG2500-series" added by "root".
Nov 23 22:05:22 Vortex cupsd[1315]: REQUEST localhost - root "POST /admin/ HTTP/1.1" 200 345 CUPS-Add-Modify-Printer successful-ok
Nov 23 22:05:22 Vortex cupsd[1315]: Printer "MG2500-series" started by "root".
Nov 23 22:05:22 Vortex cupsd[1315]: REQUEST localhost - root "POST /admin/ HTTP/1.1" 200 129 Resume-Printer successful-ok
Nov 23 22:05:22 Vortex cupsd[1315]: Printer "MG2500-series" now accepting jobs ("root").
Nov 23 22:05:22 Vortex cupsd[1315]: REQUEST localhost - root "POST /admin/ HTTP/1.1" 200 129 CUPS-Accept-Jobs successful-ok
Nov 23 22:05:22 Vortex cupsd[1315]: Default destination set to "MG2500-series" by "root".
Nov 23 22:05:22 Vortex cupsd[1315]: REQUEST localhost - root "POST /admin/ HTTP/1.1" 200 129 CUPS-Set-Default successful-ok
Nov 23 22:05:23 Vortex cupsd[1315]: Expiring subscriptions...
Nov 23 22:05:24 Vortex cupsd[1315]: Expiring subscriptions...
Nov 23 22:05:24 Vortex cupsd[1315]: REQUEST localhost - - "POST / HTTP/1.1" 200 150 Cancel-Subscription successful-ok
Nov 23 22:05:24 Vortex cupsd[1315]: REQUEST localhost - - "POST / HTTP/1.1" 200 150 Cancel-Subscription client-error-not-found
Nov 23 22:05:25 Vortex cupsd[1315]: Expiring subscriptions...
Nov 23 22:05:49 Vortex cupsd[1315]: Saving printers.conf...
Nov 23 22:05:49 Vortex cupsd[1315]: Saving classes.conf...
Nov 23 22:05:49 Vortex cupsd[1315]: Generating printcap /etc/printcap...
Nov 23 22:05:49 Vortex cupsd[1315]: Saving subscriptions.conf...
Nov 23 22:05:49 Vortex cupsd[1315]: Expiring subscriptions...
Nov 23 22:05:50 Vortex cupsd[1315]: Expiring subscriptions...
```

### Notify maintainers
*Based on commit history for the SANE and `cupsd` modules*
@symphorien @supersandro2000 @cab404 @Profpatsch @matthewbauer

### Metadata
```console
 - system: `"x86_64-linux"`
 - host os: `Linux 5.14.16-lqx1, NixOS, 21.11 (Porcupine)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.4`
 - channels(root): `""`
 - channels(v7): `""`
 - nixpkgs: `/nix/store/2qirxpxpba09i0ndh55avywirphdgkg3-source`
```
**Printer:** CANON Pixma MG 2570
**Driver:** CUPS+Gutenprint v5.3.4

### Maintainer information
```yaml
# A list of NixOS Modules affected by the problem
module:
services/hardware/sane.nix 
services/printing/cupsd.nix
```

**Comment by symphorien** [comment]
>  After disabling it and running nixos-rebuild switch, the printer is again immediately recognized, but obviously the scanner doesn't work.


that is extremely puzzling, as sane is enabled or disabled though an environment variable that cannot be changed by a mere nixos-rebuild switch. you need to reboot or to logout/login to effectively toggle it.

**Comment by maydayv7** [comment]
Yes. initially, in order to enable the scanning functionality, a reboot was required. But the connection to the printer was established by simply disabling the `hardware.sane` option and rebuilding

**Comment by Profpatsch** [comment]
Without lookling deeper into this, are you sure it’s not just a random coincidence? Were you setting up printing and scanning in the same session? Maybe you changed something else in-between.

**Comment by maydayv7** [comment]
Just tried it again today, same results

**Comment by tensor5** [comment]
I have the same issue.

**Comment by jeeger** [comment]
Pretty sure it's a permission problem.
With cups enabled:
```
[root@ogion:/dev/bus/usb]# fd -l
drwxr-xr-x 2 root root      120 Feb 25 13:15 ./001
crw-rw-r-- 1 root root 189,   0 Feb 25 12:59 ./001/001
crw-rw-r-- 1 root root 189,   2 Feb 25 12:59 ./001/003
crw-rw-r-- 1 root root 189,   3 Feb 25 12:59 ./001/004
crw-rw-r-- 1 root lp   189,   8 Feb 25 13:15 ./001/009
drwxr-xr-x 2 root root       60 Feb 21 21:25 ./002
crw-rw-r-- 1 root root 189, 128 Feb 25 12:59 ./002/001
```

With sane, saned and cups enabled:
```
[root@ogion:/dev/bus/usb]# fd -l
drwxr-xr-x 2 root root         120 Feb 25 13:19 ./001
crw-rw-r-- 1 root root    189,   0 Feb 25 12:59 ./001/001
crw-rw-r-- 1 root root    189,   2 Feb 25 12:59 ./001/003
crw-rw-r-- 1 root root    189,   3 Feb 25 12:59 ./001/004
crw-rw-r-- 1 root scanner 189,   9 Feb 25 13:19 ./001/010
drwxr-xr-x 2 root root          60 Feb 21 21:25 ./002
crw-rw-r-- 1 root root    189, 128 Feb 25 12:59 ./002/001
```
After a reboot, permissions are back to the first case. I guess there's a problem with sane's udev rules.

**Comment by symphorien** [comment]
Nice find!

Do I understand correctly that when the device is owned by lp scanning does not work and when it is owned by scanner printing does not work?

**Comment by jeeger** [comment]
Hello!
Yes. From  my understanding of [the source](https://github.com/NixOS/nixpkgs/blob/ab7314cb320d9cb7db0688d0f84b3c6b705a3094/pkgs/applications/graphics/sane/backends/default.nix#L85), `sane-backends` generates udev rules for every "scanner" device that changes the group permissions of `/dev/bus/usb` device nodes to "scanner". This does not work for multifunction devices, as Cups expects device nodes to be accessible via the "lp" group. I've created a quick fix for this by overriding the `sane-backends` derivation to delete its udev rules. 
Note that SANE is set up for device nodes being owned by the `lp` group (seems to be implied by [this](https://github.com/NixOS/nixpkgs/blob/ab7314cb320d9cb7db0688d0f84b3c6b705a3094/nixos/modules/services/hardware/sane.nix#L55).
I've worked around this by overriding sane backends to delete the udev rules:

```
      sane-backends = super.sane-backends.overrideAttrs (oldAttrs: rec {
        postInstall =
          with pkgs.lib;
          let
            alreadyPresent = hasInfix "# delete udev rules" oldAttrs.postInstall;
          in
            oldAttrs.postInstall + optionalString (!alreadyPresent) ''
              rm $out/etc/udev/rules.d/49-libsane.rules # delete udev rules
            '';
      });
```

This is a temporary fix, but I don't know what the "proper" way to do this would be.

**Comment by symphorien** [comment]
Ok so when the device is owned by lp, it also works with the scanner. Reading you it looks like an issue with upstream udev rules, I wonder if filing a bug upstream is warranted.

**Comment by symphorien** [comment]
Looks like https://github.com/OpenPrinting/cups/issues/314#issuecomment-1014325640 is the proper solution. @jeeger would you mind giving it a try? It's hard to test without affected hardware...

**Comment by jeeger** [comment]
Turns out the `sane-desc` command was never run, and the rules were just copied from the sane source. `sane-desc` requires a search path to find description files. After patching `sane-backends` via 
```nix
      sane-backends = super.sane-backends.overrideAttrs (oldAttrs: rec {
        postInstall = builtins.replaceStrings
          ["./tools/sane-desc -m udev"]
          ["./tools/sane-desc -m udev+hwdb -s doc/descriptions:doc/descriptions-external"]
          oldAttrs.postInstall;
      });
```
rules are indeed generated, and the owner isn't changed.

**Comment by symphorien** [comment]
Sounds nice. Would you mind opening a PR?

**Comment by 7FM** [comment]
> Sounds nice. Would you mind opening a PR?

@symphorien I went ahead and created https://github.com/NixOS/nixpkgs/pull/176412, due to lack of activity.

**Comment by symphorien** [comment]
this was fixed, but incorrectly, in #176412. A fix to the fix is in #185700, it would be nice if people affected by this can test it.

**Comment by 7FM** [comment]
> this was fixed, but incorrectly, in #176412. A fix to the fix is in #185700, it would be nice if people affected by this can test it.

Still works for me, tested printing & scanning with the following overlay:
```
sane-backends = prev.sane-backends.overrideAttrs (oldAttrs: rec {
  postInstall = builtins.replaceStrings
    ["mkdir -p $out/etc/udev/rules.d/"
     "./tools/sane-desc -m udev+hwdb -s doc/descriptions:doc/descriptions-external > $out/etc/udev/rules.d/49-libsane.rules || \\"
     "cp tools/udev/libsane.rules $out/etc/udev/rules.d/49-libsane.rules"]
    ["mkdir -p $out/etc/udev/rules.d/ $out/etc/udev/hwdb.d"
     "./tools/sane-desc -m udev+hwdb -s doc/descriptions:doc/descriptions-external > $out/etc/udev/rules.d/49-libsane.rules"
     "./tools/sane-desc -m udev+hwdb -s doc/descriptions -m hwdb > $out/etc/udev/hwdb.d/20-sane.hwdb"]
    oldAttrs.postInstall;
});
```

**Comment by abathur** [comment]
Closing since it sounds like some combination of https://github.com/NixOS/nixpkgs/pull/185700 and https://github.com/NixOS/nixpkgs/pull/199550 fixed this, but I'm happy to reopen if I'm misunderstanding.
